torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -895,6 +911,21 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
 
 void ProcessGroupNCCL::ncclCommWatchdog() {
   try {
+    // Inside watchdog thread, we might still call CUDA Event and it will
+    // hang the thread and stuck there silent for hours. We want to use a
+    // timer to kill the process if watchdog timeout.
+    struct sigevent action;
+    std::memset(&action, 0, sizeof(action));
+    action.sigev_notify = SIGEV_THREAD;
+    action.sigev_notify_function = killNCCLPGWatchDogDeadlock;
+    int rv = ::timer_create(CLOCK_REALTIME, &action, &timerId_);
```

Comment:
probably use CLOCK_MONOTONIC here? realtime clock can be 'set' and for watchdog purposes we just want reliable, relative time